ow_sim_tests/test/terrain_interaction.py

```diff
@@ -0,0 +1,329 @@
+      "\n===%s action completed in %0.3fs===" % (action_name, elapsed)
+    )
+
+    # verify action ended where we expected
+    self._assert_point_is_near(
+      result.final, expected_final, 0.02,
+      "Arm did not complete the %s action in the position expected!"
+        % action_name
+    )
+
+    return result
+
+  """
+  Tests an extra deep grind action and asserts no regolith is spawned during the
+  operation.
+  """
+  def test_01_grind(self):
+
+    GRIND_MAX_DURATION = 80.0 # seconds
+    GRIND_EXPECTED_FINAL = Point(1.4720579324199088,
+                                 -0.14069361533007319,
+                                 -6.739956216168468)
+
+    # call Grind action asynchronously
+    grind_result = self._test_action(
+      'Grind',
+      ow_lander.msg.GrindAction,
+      ow_lander.msg.GrindGoal(),
+      GRIND_MAX_DURATION,
+      GRIND_EXPECTED_FINAL,
+      self._assert_regolith_not_present,
+      server_timeout = 20.0, # (seconds) first action call needs longer timeout
+      x_start         = 1.65,
+      y_start         = 0.0,
+      depth           = 0.15, # grind deep so terrain is modified
+      length          = 0.7,
+      parallel        = True,
+      ground_position = GROUND_POSITION
+    )
+
+  """
+  Tests a default dig linear action and asserts regolith does spawn during the
+  operation and remains in the scoop until the end of it.
+  """
+  def test_02_dig_linear(self):
+
+    DIG_LINEAR_MAX_DURATION = 110.0
+    DIG_LINEAR_EXPECTED_FINAL = Point(2.2404188541487606,
+                                      -0.012052180209644802,
+                                      -7.049335444887187)
+
+    # call Grind action asynchronously
+    dig_linear_result = self._test_action(
+      'DigLinear',
+      ow_lander.msg.DigLinearAction,
+      ow_lander.msg.DigLinearGoal(),
+      DIG_LINEAR_MAX_DURATION,
+      DIG_LINEAR_EXPECTED_FINAL,
+      self._assert_regolith_is_in_scoop,
+      x_start         = 1.46,
+      y_start         = 0.0,
+      depth           = 0.01,
+      length          = 0.1,
+      ground_position = GROUND_POSITION
+    )
+
+    # verify regolith models spawned
+    self._assert_regolith_present()
+
+  """
+  Tests a deliver action and asserts that regolith remains in the scoop until
+  the final portion when it is dumped into the sample dock. Upon completion of
+  the operation it also asserts that regolith models are deleted.
+  """
+  def test_03_deliver(self):
+
+    DELIVER_MAX_DURATION = 60.0 # TODO: constrain
+    DELIVER_EXPECTED_FINAL = Point(0.5562290134759807,
+                                   -0.21354780470240303,
+                                   -6.351096861727068)
+
+    REGOLITH_CLEANUP_WAIT = 5.0 # seconds
+
+    deliver_result = self._test_action(
+      'Deliver',
+      ow_lander.msg.DeliverAction,
+      ow_lander.msg.DeliverGoal(),
+      DELIVER_MAX_DURATION,
+      DELIVER_EXPECTED_FINAL,
+      self._assert_regolith_transports_and_delivers
+    )
+
```

Comment:
#suggestion: Do you think that you can add a check that verifies that the samples were off-loaded of the scoop based on their proximity to the scoop?

Reply:
For example if you refactor `_assert_regolith_is_in_scoop` to check if near or if far then it could be easy to use here for this check